redbot/cogs/mutes/mutes.py

```diff
@@ -0,0 +1,782 @@
+                if not success:
+                    mute_success.append(f"{channel.mention} - {issue}")
+                await asyncio.sleep(0.1)
+            if mute_success and len(mute_success) == len(guild.channels):
+                return False, "\n".join(s for s in mute_success)
+            elif mute_success and len(mute_success) != len(guild.channels):
+                return True, "\n".join(s for s in mute_success)
+            else:
+                return True, None
+
+    async def unmute_user(
+        self, guild: discord.Guild, author: discord.Member, user: discord.Member, reason: str,
+    ) -> Tuple[bool, Optional[str]]:
+        """
+            Handles muting users
+        """
+        mute_role = await self.config.guild(guild).mute_role()
+        if mute_role:
+            try:
+                if not await is_allowed_by_hierarchy(self.bot, self.config, guild, author, user):
+                    return False, _(mute_unmute_issues["hierarchy_problem"])
+                role = guild.get_role(mute_role)
+                if not role:
+                    return False, mute_unmute_issues["role_missing"]
+                await user.remove_roles(role, reason=reason)
+            except discord.errors.Forbidden:
+                return False, mute_unmute_issues["permissions_issue"]
+            return True, None
+        else:
+            mute_success = []
+            for channel in guild.channels:
+                success, issue = await self.channel_unmute_user(
+                    guild, channel, author, user, reason
+                )
+                if not success:
+                    mute_success.append(f"{channel.mention} - {issue}")
+                await asyncio.sleep(0.1)
+            if mute_success:
+                return False, "\n".join(s for s in mute_success)
+            else:
+                return True, None
+
+    async def channel_mute_user(
+        self,
+        guild: discord.Guild,
+        channel: discord.abc.GuildChannel,
+        author: discord.Member,
+        user: discord.Member,
+        reason: str,
+    ) -> Tuple[bool, Optional[str]]:
+        """Mutes the specified user in the specified channel"""
+        overwrites = channel.overwrites_for(user)
+        permissions = channel.permissions_for(user)
+
+        if permissions.administrator:
+            return False, _(mute_unmute_issues["is_admin"])
+
+        new_overs = {}
+        if not isinstance(channel, discord.TextChannel):
+            new_overs.update(speak=False)
+        if not isinstance(channel, discord.VoiceChannel):
+            new_overs.update(send_messages=False, add_reactions=False)
+
+        if not await is_allowed_by_hierarchy(self.bot, self.config, guild, author, user):
+            return False, _(mute_unmute_issues["hierarchy_problem"])
+
+        old_overs = {k: getattr(overwrites, k) for k in new_overs}
+        overwrites.update(**new_overs)
```

Comment:
Storing the old overwrites is an error that could change other permissions upon unmute, this should only store the difference in permissions caused by the mute.

Reply:
This only stores the old overwrites for specific permissions on that user. `overwrites` above is the channel perms for that user only and the state they were in before they were muted. This was straight from the old mutes in mod.